### DISCUSSION TITLE
Usb CDC, VENDOR and HID at the same time

### DISCUSSION THREAD

**Discussion by TheStaticTurtle** [hathach/tinyusb#1202] [Q&A] [answered]
I'm currently working on a project based on an RP2040. The device would be configured by CDC or WebUSB and would act as an HID keyboard.

<details>
  <summary>Original issue</summary>
  

> I've been working on this for hours and I can't make it work. After I add the HID interface to the existing CDC and VENDOR, windows says that it can't bring the device up and Linux says `can't set config #1, error -32`. (The CDC and VENDOR were working fine)
>   
>   Here is my tinyusb config code:
>   https://gist.github.com/TheStaticTurtle/47207c8e5f19512315ec37099aecf5b9
>   
>   EDIT: USB Device tree viewer output: https://pastebin.com/8MU0UGSJ
>   
>   I don't think I missed something, so is there a limit on how many interfaces you can have? Or are some interface incompatible?

</details>

So as @hathach requested, I forked the repo and add an example https://github.com/TheStaticTurtle/tinyusb/tree/example-webusb-cdc-hid/examples/device/webusb_cdc_hid
For compilation, I use the pico-example build script (CMake)

So there is some improvement, as far as I can tell, I've now got all the interface showing up and recognized corretly ([USB Device tree log](https://pastebin.com/xSqZDddy)), CDC and WebUSB seem to be somewhat working (sometimes needs an unplug / re-plug).
But HID (Keyboard or Mouse) is nowhere to be seen (One theory I have is that the board_button_read doesn't actually work as expected with the BOOTSEL button on the raspberry pico (Seems to hang everything when pressed)). The hid callback also doesn't work.

**Comment by hathach** [comment, marked as answer]
the code is preferably compiled like one in example folder and re-viewable for people to comment on. The easiest way is
- Fork this repo, checkout a new branch
- Add your-own-example based on stock one
- Push and post it here.

**Reply by TheStaticTurtle** [reply]
Done, see my updated post 😃

**Reply by hathach** [reply]
> But HID (Keyboard or Mouse) is nowhere to be seen (One theory I have is that the board_button_read doesn't actually work as expected with the BOOTSEL button on the raspberry pico (Seems to hang everything when pressed)). The hid callback also doesn't work.

bootsel button works fine, you can try with stock example `hid_composite` to verify the hardware.

The issue is with the send report API
https://github.com/TheStaticTurtle/tinyusb/blob/example-webusb-cdc-hid/examples/device/webusb_cdc_hid/src/main.c#L200

all class driver API take instance ID instead of interface number. Which is zero-index and depends on the order of appearance of the interfaces within configuration. I should rename the ITF to INSTANCE to make it clearer.

https://github.com/hathach/tinyusb/blob/master/examples/device/hid_multiple_interface/src/main.c#L37

PS: Maybe I should change it to use actual interface number, though it would be an unnecessary breaking changes.

**Reply by TheStaticTurtle** [reply]
Ok I got it working on my example I'll push it to my repo later, do you want a pr for that example ?
> bootsel button works fine, you can try with stock example hid_composite to verify the hardware.

True, that's mybad, I thought it might have some issues because it acts on the flash pin but I was wrong 😅 

So can you clear this up for me: 
As I understand it, you can have multiple interfaces (cdc, vendor, hid, hid), but you actually need to use the instance number of that interface (ie cdc0, vendor0, hid0, hid1 in my project).
So for an other example this would also be right ??: cdc0, cdc1, hid0, vendor0, hid1

Thanks a lot 😄

**Reply by hathach** [reply]
> As I understand it, you can have multiple interfaces (cdc, vendor, hid, hid), but you actually need to use the instance number of that interface (ie cdc0, vendor0, hid0, hid1 in my project).

no problem, `tud_class()` is default to use instance0 which is 90% of the user usage. `tud_class_n()` take extra instance arg, zero index, consecutive number. Yeah it is cdc0, cdc1, hid0, vendor0, hid1 and not interface number. If it works, don't forget to mark it as anwsered.

**Reply by TheStaticTurtle** [reply]
Thanks a lot 😄

**Comment by thewh1teagle** [comment]
@TheStaticTurtle 
Thank you so much for this example. I'm trying few days to use CDC + HID and this is the first example I found which working (your fork)

**Reply by ihab** [reply]
Hi there! I am using TinyUSB on the ESP32-S3 via the ESP-IDF provided by Espressif. I have the following question that I posted to the ESP32 forums:

https://esp32.com/viewtopic.php?f=13&t=37912&p=126519#p126519

This is with --

`espressif/esp_tinyusb` 1.4.3 https://components.espressif.com/components/espressif/esp_tinyusb
`espressif/tinyusb` 0.15.0~4 https://components.espressif.com/components/espressif/tinyusb

What is known about Espressif's fork of TinyUSB? Is that fork capable of supporting a CDC and a HID interface simultaneously?
